### PR TITLE
Feature trading

### PR DESCRIPTION
**Pull request by jastornig**

### CODE REVIEW

File: src/main/java/com/group2/catanbackend/dto/game/AcceptMoveDto.java
Comment (l1wobak) [inline src/main/java/com/group2/catanbackend/dto/game/AcceptMoveDto.java:1]: See frontend comments

File: src/main/java/com/group2/catanbackend/dto/game/TradeOfferDto.java
Comment (l1wobak) [inline src/main/java/com/group2/catanbackend/dto/game/TradeOfferDto.java:1]: See frontend comments

File: src/main/java/com/group2/catanbackend/dto/game/TradeOfferDto.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        this.fromPlayer = fromPlayer;
+        this.setEventType(MessageType.PLAYER_NOTIFY);
+    }
+    //order swapped!!!
+    private int[] getResources;
+    private int[] giveResources;
+    private IngamePlayerDto fromPlayer;
+
+
+    public boolean equals(TradeOfferDto compare){
```
Comment (l1wobak) [inline src/main/java/com/group2/catanbackend/dto/game/TradeOfferDto.java]: Fix sonarcloud issue!!!!!!!!

File: src/main/java/com/group2/catanbackend/gamelogic/GameLogicController.java [hunk 2]
```diff
@@ -163,6 +174,99 @@ private boolean stealResource(Player playerToStealFrom, Player playerToGiveTo) {
...
+        sendCurrentGameStateToPlayers();
+        //send new GameProgressDto? but with what content?
+    }
+
+    //TODO: rethink checks. I did not until now
+    private void makeTradeMove(TradeMoveDto tradeMove, Player player){
+        //if (isSetupPhase) throw new InvalidGameMoveException(ErrorCode.ERROR_CANT_ROLL_IN_SETUP);
+        if (activePlayer != player)
+            throw new NotActivePlayerException(ErrorCode.ERROR_NOT_ACTIVE_PLAYER.formatted(activePlayer.getDisplayName()));
+        if(tradeMove.getToPlayers().size() != players.size()-1)
```
Comment (l1wobak) [inline src/main/java/com/group2/catanbackend/gamelogic/GameLogicController.java]: doesnt this always fail if: 
- you dont want to send the trade to every player
- you want to trade with the bank
as then the getToPlayers list is not the same size as the players list -1?

File: src/main/java/com/group2/catanbackend/gamelogic/GameLogicController.java [hunk 3]
```diff
@@ -163,6 +174,99 @@ private boolean stealResource(Player playerToStealFrom, Player playerToGiveTo) {
...
+            throw new NotActivePlayerException(ErrorCode.ERROR_INVALID_CONFIGURATION);
+        if (!player.resourcesSufficient(tradeMove.getGiveResources()))
+            throw new InvalidGameMoveException(ErrorCode.ERROR_NOT_ENOUGH_RESOURCES.formatted(tradeMove.getClass().getSimpleName()));
+        /*
+        if(tradeMove.getWaitTime()<1)
+            throw new InvalidConfigurationException(ErrorCode.ERROR_INVALID_CONFIGURATION);
+         */
+        if(tradeMove.getToPlayers().isEmpty() ||isempty(tradeMove.getToPlayers())){
```
Comment (l1wobak) [inline src/main/java/com/group2/catanbackend/gamelogic/GameLogicController.java]: why check the same thing twice? also, i think this is unreachable as you throw an error before when checking if toPlayersList.size != players.size which is only possible to be true if there are exactly 2 players.

**Review by l1wobak — COMMENTED** [review]
Write Testcases!!!!!!!!!!

**Comment by jastornig** [comment]
Hier dasselbe wie in android
Meine letzten commits waren nur fürs bauen eines docker containers, welcher übrigens jetzt am uni server erreichbar ist. 

@Ghost-Element i empfehl dir wirklich, die da ran zu setzen. Mittwoch Vormittag is zu spät dafür.
Todos im code beachten.
Feedback von den reviewern
Tests tests tests.
Sonar issues. 

https://sonarcloud.io/summary/new_code?id=SE-II-group2_catan-backend&pullRequest=33

**Review by szolderits — APPROVED** [review]

**Comment by jastornig** [comment]
As no response was received from @Ghost-Element we decided to continue developing the trading feature to a point, where it seems to be stable and does not crash the app anymore. We will merge this PR without reaching the sonar coverage. It will be left untested.

**Pull request merged**